### PR TITLE
Metrics for search latencies

### PR DESCRIPTION
**Pull request by JVerwolf**
This PR adds metrics for search transport latencies, expressed as percentiles.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.search;
+
+import org.elasticsearch.telemetry.TelemetryProvider;
+import org.elasticsearch.telemetry.metric.LongHistogram;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+
+public class SearchTransportAPMMetrics {
```
Comment (JVerwolf) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:15]: This comment explains the purpose of this class: https://github.com/elastic/elasticsearch/pull/101577/files#r1379132909

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    private final LongHistogram actionLatencies;
+
+    public SearchTransportAPMMetrics(MeterRegistry meterRegistry) {
+        this(
+            meterRegistry.registerLongHistogram(
+                "es.transport.action.latency.histogram",
+                "Transport action execution times expressed as a histogram",
+                "millis"
+            )
```
Comment (JVerwolf) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:24]: A histogram metric type dynamically creates buckets based on the data it receives. 

Histograms (e.g. percentiles) must be calculated at the point of collection as opposed to trying to create one in Kibana that is derived from a single counter metric. This is because timing data is discarded when using a single counter, as all recorded measurements are stored in an single Atomic Long (or equivalent), and timing is only ascribed when the metrics are polled by the collection pipeline.  Thus, if we need that timing granularity to accurately produce percentiles, we must record it in a histogram at the point of collection.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java [hunk 3]
```diff
@@ -507,11 +511,32 @@ public static void registerRequestHandler(TransportService transportService, Sea
...
+        Writeable.Reader<Request> requestReader,
+        TransportRequestHandler<Request> handler
+    ) {
+        transportService.registerRequestHandler(action, executor, requestReader, (request, channel, task) -> {
+            var startTime = transportService.getThreadPool().relativeTimeInMillis();
+            try {
+                handler.messageReceived(request, channel, task);
+            } finally {
+                var elapsedTime = transportService.getThreadPool().relativeTimeInMillis() - startTime;
+                searchTransportMetrics.getActionLatencies().record(elapsedTime, Map.of("action", action));
```
Comment (JVerwolf) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:529]: The map `Map.of("action", action)` passed into the parameter here is used for "attributes" (similar to "labels" or "tags" in other popular metric libraries). Attributes are the dynamic portion of the metric "key" (name) that can be used for filtering and aggregation.  They must be low cardinality.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java [hunk 4]
```diff
@@ -465,7 +469,7 @@ public static void registerRequestHandler(TransportService transportService, Sea
         );
         TransportActionProxy.registerProxyAction(transportService, QUERY_SCROLL_ACTION_NAME, true, ScrollQuerySearchResult::new);
 
-        transportService.registerRequestHandler(
+        registerRequestHandlerWithAPMMetrics(
```
Comment (JVerwolf) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:472]: Hi @javanna , I'm a bit unsure as to what some of these phases/actions do, can you please check that I've instrumented the right ones here? Thanks!

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+package org.elasticsearch.action.search;
+
+import org.elasticsearch.telemetry.TelemetryProvider;
+import org.elasticsearch.telemetry.metric.LongHistogram;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+
+public class SearchTransportAPMMetrics {
+    private final LongHistogram actionLatencies;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:16]: Question: we are going to store latencies for different phases using this same variable, but with different attributes? Intuitively, and maybe naively, i would have expected one variable per type of latency.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+    public SearchTransportAPMMetrics(MeterRegistry meterRegistry) {
+        this(
+            meterRegistry.registerLongHistogram(
+                "es.transport.action.latency.histogram",
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:21]: The class is about search but the metric does not include search, is that something that we should fix?

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                "Transport action execution times expressed as a histogram",
+                "millis"
+            )
+        );
+    }
+
+    public SearchTransportAPMMetrics(LongHistogram actionLatencies) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:28]: This could be private?

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java (hunk 4 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:472]: we may want to instrument FETCH_ID and CAN_MATCH as well, but before we do that, what is the purpose of this instrumentation? 

This traces how long it takes to execute the phase on each shard. That is very granular and probably useful. Another place to start, or possibly both are good, is to track how long each phase takes overall, from a coordinator perspective. What do you think?

Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:472]: Thanks for the feedback! 
> We may want to instrument FETCH_ID and CAN_MATCH

Can do!

> Another place to start, or possibly both are good, is to track how long each phase takes overall, from a coordinator perspective.

Ah, good point.  I think I started here because, according to my notes, we discussed adding instrumentation here.   

Thinking about the current implementation a bit - the stats from each individual node will all collectively contribute to the same metric in Kibana, so one can still determine phase latencies/performance using instrumentation at this location.  

However, a there is a difference that comes to mind wrt collecting a similar metric at the coordinator.  When the phase is fanned out and collected, the overall phase time on the coordinator is that of the longest individual shard's phase, plus networking and coordination.  

If there exists an underlying issue whereby a single shard runs much longer than the others (for a given phase), the current metric will be able to expose that whereas an overall metric (from the coordinator's perspective) won't.  Theoretically, one could monitor the p99 (or equivalent) bucket in the histogram to try to detect these issues and use metric attributes to determine if the issue is common to the node/shard, or distributed across the cluster.  This could be useful information, at least in my mind 🙂 

I do think that the overall phase latency from the coordinator's perspective may be interesting as well in identifying overall trends and potentially answering other questions, but perhaps not enough on it's own.  At least that's my current thinking.  WDYT?

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java (hunk 6 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:21]: Good point.  I'm thinking we'll later add another metric at the coordinator,  like so:
`es.search.phase.coordinator.latency.histogram`

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java (hunk 5 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:16]: They way I think about it (and I could be wrong), is that each of the possible combinations  of the "static" name and the "dynamic" attributes can be considered a separate time-series.  The attributes are useful when performing aggregations.  I was thinking it could be the case that we'd want to group various phases together in some of our reporting (with the option of separating them when drilling down), but I could be mistaken 🙂  WDYT?
 
 The [proposal](https://docs.google.com/document/d/1jKxuaZi7QAMIRD_Eq3nonkYswVlQXW5TWllJEicoOtM/edit) mentions:
- When there are different "flavors" of a metric (i.e. s3, azure, etc) use an attribute rather than inserting it in the metric name.
- Rule of thumb: you should be able to do aggregations (e.g. sum, avg) across a dimension of a given metric (without the need to aggregate over different metric names); on the other hand, any aggregation across any dimension of a given metric should be meaningful.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:16]: I find this hard to follow. Do you have an example of pros and cons of the two approaches (single metric vs multiple metrics)?

Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:16]: I was thinking we'd want to monitor/chart transport action latencies in general (or at least a subset of them that we care about). 

In any of these latency metrics, the value at any given time matters little. What  matters (IMO) is trends: changes in metric values in aggregate over time.  If we see transport actions latencies increase dramatically for a given node, we know there's a problem.  If we see transport actions that are more expensive across the board for a given cluster/cloud-provider/region/version/etc we know there's a problem. We can gain that insight more easily if we have a single metric that we monitor, rather than manually defining the chart/alert as a collection of metrics which we must remember to update if the code changes.  

I don't think we loose anything by labeling the transport actions as attributes on the metric, as we can always filter for specific actions to "drill down".

At least, that's my understanding. Much of my metrics understanding was learned from using other time-series databases, so  I'm happy to be persuaded otherwise though if you're thinking of something I'm missing!

Metrics can be thought of as an API, and APIs are hard to change. I think we should have a conversation about this to ensure we get this right.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:16]: As far as I understand there isn't a huge difference between single metric and multiple metrics. I am fine with using one with attributes.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:15]: I can see why we ended up here. It currently feels a bit weird to have one class that includes this single instance member. Why isn't the instance member directly added to the only consumer of this new class? perhaps we are planning on adding more later? Not a big deal for me, just trying to understand the choice.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+                "millis"
+            )
+        );
+    }
+
+    private SearchTransportAPMMetrics(LongHistogram actionLatencies) {
+        this.actionLatencies = actionLatencies;
+    }
+
+    public static SearchTransportAPMMetrics NOOP = new SearchTransportAPMMetrics(TelemetryProvider.NOOP.getMeterRegistry());
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:32]: Is this used only in tests? If so it could be created on the fly in tests or live on the test side rather than in the prod code.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java [hunk 9]
```diff
@@ -382,7 +384,7 @@ public void writeTo(StreamOutput out) throws IOException {
         }
     }
 
-    public static void registerRequestHandler(TransportService transportService, SearchService searchService) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:385]: I think that you could leave this static and add the SearchTransportMetrics argument, not a huge deal but it helps following dependencies.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java [hunk 10]
```diff
@@ -91,11 +91,13 @@ public SearchTransportService(
...
             Transport.Connection,
             SearchActionListener<? super SearchPhaseResult>,
-            ActionListener<? super SearchPhaseResult>> responseWrapper
+            ActionListener<? super SearchPhaseResult>> responseWrapper,
+        SearchTransportAPMMetrics searchTransportMetrics
     ) {
         this.transportService = transportService;
         this.client = client;
         this.responseWrapper = responseWrapper;
+        this.searchTransportMetrics = searchTransportMetrics;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:100]: should we add a test that verifies that the instrumentation actually works and the metrics can be retrieved?

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java [hunk 11]
```diff
@@ -382,7 +384,7 @@ public void writeTo(StreamOutput out) throws IOException {
         }
     }
 
-    public static void registerRequestHandler(TransportService transportService, SearchService searchService) {
+    public void registerRequestHandler(TransportService transportService, SearchService searchService) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:387]: Proposal: shall we instrument all of the handlers registered in this class, excluding the proxy ones? That would mean instrumenting clear scroll and free context too? I think that would not hurt.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java [hunk 12]
```diff
@@ -423,7 +425,10 @@ public static void registerRequestHandler(TransportService transportService, Sea
             DFS_ACTION_NAME,
             EsExecutors.DIRECT_EXECUTOR_SERVICE,
             ShardSearchRequest::new,
-            (request, channel, task) -> searchService.executeDfsPhase(request, (SearchShardTask) task, new ChannelActionListener<>(channel))
+            (request, channel, task) -> instrumentedHandler(
+                DFS_ACTION_NAME,
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:429]: I just realized that the overall notion that these are measured on the shard, is not entirely accurate. can match is a case where we group shard requests per data node. Hence the timing here would be "how long does it take for can match to be executed on a certain data node". It is still a good metric, but it covers more than one shard. For individual can match on each shard, we need to use a different mechanism. Perhaps we need to adjust the name of the metric then and keep it more general. We could also reuse it for coord level timing, and perhaps the action name needs to include the relevant context?

Also, I would use a more readable action name for our metrics. Some of these action names are hard to distinguish from one another. It took me a bit to trace back all the relevant info, hence I am adding some comments around each action.

This is the can match phase which is executed to pre filter shards that a search request hits. It rewrites the query on the shard and checks whether the result of the rewrite matches no documents, in which case the shard can be filtered out:

- QUERY_CAN_MATCH_NODE_NAME = "shards_can_match";

These two are specific to DFS_QUERY_THEN_FETCH that fetches distributed term frequencies, and executes KNN (the top level search element):
- DFS_ACTION_NAME: dfs_query_then_fetch/shard_dfs_phase
- QUERY_ID_ACTION_NAME = dfs_query_then_fetch/shard_query_phase

This it the plain old query phase if no DFS and KNN or scroll are requested:
- QUERY_ACTION_NAME = query_then_fetch/shard_query_phase

- FREE_CONTEXT_ACTION_NAME: shard_release_context
- 
This one is shared by query_then_fetch and dfs_query_then_fetch
- FETCH_ID_ACTION_NAME: shard_fetch_phase


These three are scroll specific: query then fetch, or a shortcut scenario where query and fetch are done in the shard as part of a single roundtrip:
- QUERY_SCROLL_ACTION_NAME: scroll/shard_query_phase
- FETCH_ID_SCROLL_ACTION_NAME = scroll/shard_fetch_phase
- QUERY_FETCH_SCROLL_ACTION_NAME = scroll/shard_query_and_fetch_phase
- FREE_CONTEXT_SCROLL_ACTION_NAME: scroll/shard_release_context
- CLEAR_SCROLL_CONTEXTS_ACTION_NAME: scroll/shard_release_contexts



I would also add metrics for the coordinator part of these phases:

- coord/can_match
- dfs_query_then_fetch/coord_dfs_phase
- dfs_query_then_fetch/coord_query_phase
- query_then_fetch/coord_query_phase
- coord_fetch_phase
- scroll/coord_query_phase
- scroll/coord_fetch_phase
- scroll/coord_query_and_fetch_phase
- scroll/coord_release_context
- scroll/coord_release_contexts
- coord_release_context


Does the naming make sense? Do the additional coord metrics sound good to you?

**Review by javanna — COMMENTED** [review]
I left a few comments, looks good overall.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java (hunk 9 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:385]: I spent some time pulling on this thread; the changes are contained in this commit: https://github.com/elastic/elasticsearch/pull/102557/commits/40a878bef230a509fce67ef1033d6d1c8071287f

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java [hunk 13]
```diff
@@ -494,24 +499,41 @@ public static void registerRequestHandler(TransportService transportService, Sea
...
 
+    private void instrumentedHandler(String actionName, Runnable runnable) {
+        var startTime = transportService.getThreadPool().relativeTimeInMillis();
+        try {
+            runnable.run();
+        } finally {
+            var elapsedTime = transportService.getThreadPool().relativeTimeInMillis() - startTime;
+            searchTransportMetrics.getActionLatencies().record(elapsedTime, Map.of("action", actionName));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:583]: Agreed, this is cleaner 🙂 

Updated in: https://github.com/elastic/elasticsearch/pull/102557/commits/6d4f74dced10a29932dfc24c834df54ab547f33f

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java (hunk 11 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java]: Can do, added as part of https://github.com/elastic/elasticsearch/pull/102557/commits/76ed555a39d08000fcb27e914efe3257d2888b76

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java (hunk 12 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:433]: Excellent comment, thanks for the time you took to write this!

>I just realized that the overall notion that these are measured on the shard, is not entirely accurate. can match is a case where we group shard requests per data node.

To this effect, should I remove the `shard_*` references in your descriptions then? e.g. ` "shards_can_match"` -> ` "can_match"`

File: server/src/main/java/org/elasticsearch/node/NodeConstruction.java [hunk 14]
```diff
@@ -1038,6 +1040,7 @@ record PluginServiceInstances(
             b.bind(MetadataCreateIndexService.class).toInstance(metadataCreateIndexService);
             b.bind(MetadataUpdateSettingsService.class).toInstance(metadataUpdateSettingsService);
             b.bind(SearchService.class).toInstance(searchService);
+            b.bind(SearchTransportAPMMetrics.class).toInstance(searchTransportAPMMetrics);
```
Comment (piergm) [inline server/src/main/java/org/elasticsearch/node/NodeConstruction.java:1043]: PR looks really good! I have only one question and it's about this line. I don't quite understand why are we add this here, is because we want the instance injected in the `TransportSeachAction` or am I missing something?

Comment (javanna) [reply server/src/main/java/org/elasticsearch/node/NodeConstruction.java:1043]: Yep that is what binds the component and makes it available to those classes that use dependency injection (like transport actions)

Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/node/NodeConstruction.java:1043]: If we (re)make the registration static, as requested by @javanna [in this comment](https://github.com/elastic/elasticsearch/pull/102557#discussion_r1415376861), then we can no longer access `searchTransportAPMMetrics` as an instance member  of `TransportSeachAction` and instead need to provide it through dependency injection.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java (hunk 12 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java]: Ah, I missunderstood.  Reverted.

**Review by piergm — APPROVED** [review]
LGTM, thanks for the iterations.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    public static final String DFS_ACTION_METRIC = "dfs_query_then_fetch/shard_dfs_phase";
+    public static final String QUERY_ID_ACTION_METRIC = "dfs_query_then_fetch/shard_query_phase";
+    public static final String QUERY_ACTION_METRIC = "query_then_fetch/shard_query_phase";
+    public static final String FREE_CONTEXT_ACTION_METRIC = "shard_release_context"; // TODO: Not sure how to test.
+    public static final String FETCH_ID_ACTION_METRIC = "shard_fetch_phase";
+    public static final String QUERY_SCROLL_ACTION_METRIC = "scroll/shard_query_phase";
+    public static final String FETCH_ID_SCROLL_ACTION_METRIC = "scroll/shard_fetch_phase";
+    public static final String QUERY_FETCH_SCROLL_ACTION_METRIC = "scroll/shard_query_and_fetch_phase"; // TODO: Not sure how to test.
+    public static final String FREE_CONTEXT_SCROLL_ACTION_METRIC = "scroll/shard_release_context";
+    public static final String CLEAR_SCROLL_CONTEXTS_ACTION_METRIC = "scroll/shard_release_contexts"; // TODO: Not sure how to test.
```
Comment (JVerwolf) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:28]: Hey @javanna , I'm not sure how to trigger the phases with the `TODO` comments in the tests.   The can match phase seems to be triggered intermittently. Any ideas?

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java (hunk 1 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:14]: It _is_ weird I agree.  The root of the issue is that if you double register a metric, it will throw an exception.  This is not an APM thing, but a choice made/enforced by the folks to implemented our library on APM. Otherwise it would have been safe to just register a metric in a constructor then store a reference to it in a member variable, as you describe.  

So instead, we need to ensure that the lifecycle of the metric is tied to the lifecycle of the node / MeterRegistry.  One way of doing that it to hold the metric inside a wrapper class that guarantees the metric is already registered, then pass that wrapper into other spots where the lifecycle is not tied to the node's (or is but could change in future).  It's a lot of boilerplate and I don't love it, but it's better than exceptions that only present themselves under class-reloading/second-instantiation at runtime.

WDYT?

**Comment by JVerwolf** [comment]
@elasticmachine run bwc-snapshots

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java (hunk 15 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:28]: I'm going to remove the TODO comments, these are the ones I'm unsure how to test:
```
public static final String QUERY_CAN_MATCH_NODE_METRIC = "shards_can_match";
public static final String FREE_CONTEXT_ACTION_METRIC = "shard_release_context";
public static final String QUERY_FETCH_SCROLL_ACTION_METRIC = "scroll/shard_query_and_fetch_phase";
public static final String CLEAR_SCROLL_CONTEXTS_ACTION_METRIC = "scroll/shard_release_contexts";
```

**Review by piergm — APPROVED** [review]
LGTM!

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java (hunk 1 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:14]: Put another way, this purpose of this class is not for organization but instead to help enforce correctness.

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    public static final String DFS_ACTION_METRIC = "dfs_query_then_fetch/shard_dfs_phase";
+    public static final String QUERY_ID_ACTION_METRIC = "dfs_query_then_fetch/shard_query_phase";
+    public static final String QUERY_ACTION_METRIC = "query_then_fetch/shard_query_phase";
+    public static final String FREE_CONTEXT_ACTION_METRIC = "shard_release_context";
+    public static final String FETCH_ID_ACTION_METRIC = "shard_fetch_phase";
+    public static final String QUERY_SCROLL_ACTION_METRIC = "scroll/shard_query_phase";
+    public static final String FETCH_ID_SCROLL_ACTION_METRIC = "scroll/shard_fetch_phase";
+    public static final String QUERY_FETCH_SCROLL_ACTION_METRIC = "scroll/shard_query_and_fetch_phase";
+    public static final String FREE_CONTEXT_SCROLL_ACTION_METRIC = "scroll/shard_release_context";
+    public static final String CLEAR_SCROLL_CONTEXTS_ACTION_METRIC = "scroll/shard_release_contexts";
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportAPMMetrics.java:28]: Why are we still using these cryptic names in constant names? :) Shall we rename them and make them more self-explanatory?

File: server/src/test/java/org/elasticsearch/search/TelemetryMetrics/SearchTransportTelemetryTests.java [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+import static org.elasticsearch.action.search.SearchTransportAPMMetrics.FREE_CONTEXT_SCROLL_ACTION_METRIC;
+import static org.elasticsearch.action.search.SearchTransportAPMMetrics.QUERY_ACTION_METRIC;
+import static org.elasticsearch.action.search.SearchTransportAPMMetrics.QUERY_ID_ACTION_METRIC;
+import static org.elasticsearch.action.search.SearchTransportAPMMetrics.QUERY_SCROLL_ACTION_METRIC;
+import static org.elasticsearch.index.query.QueryBuilders.simpleQueryStringQuery;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertScrollResponsesAndHitCount;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertSearchHitsWithoutFailures;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 1)
+public class SearchTransportTelemetryTests extends ESIntegTestCase {
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/search/TelemetryMetrics/SearchTransportTelemetryTests.java:34]: Did we need multiple nodes for this test? Could we have used ESSingleNodeTestCase which is simpler?
